### PR TITLE
New flow controller public API [10943]

### PR DESCRIPTION
**Pull request by richiware**
Declaration of new flow controller public API for Fast-DDS.

### CODE REVIEW

**Review by rsanchez15 — CHANGES_REQUESTED** [review]
Please double check the CI jobs as there are some warnings in all of them.

File: include/fastdds/rtps/flowcontrol/FlowControllerDescriptor.hpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FASTDDS_RTPS_FLOWCONTROL_FLOWCONTROLLERDESCRIPTOR_HPP
+#define FASTDDS_RTPS_FLOWCONTROL_FLOWCONTROLLERDESCRIPTOR_HPP
+
+#include "FlowControllerDefs.hpp"
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/rtps/flowcontrol/FlowControllerDescriptor.hpp:18]: There is no coding style for this right now, and using \" in this case will be faster for the preprocessor.

File: include/fastdds/dds/domain/qos/DomainParticipantQos.hpp [hunk 2]
```diff
@@ -258,6 +260,16 @@ class DomainParticipantQos
         name_ = value;
     }
 
+    std::vector<std::shared_ptr<fastdds::rtps::FlowControllerDescriptor>>& flow_controllers()
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/domain/qos/DomainParticipantQos.hpp]: ✅

File: include/fastdds/rtps/flowcontrol/FlowControllerDefs.hpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    //! Round Robind scheduler: schedules one sample of each DataWriter in circular order.
+    ROUND_ROBLIN_SCHEDULER,
+    //! High priority scheduler: samples with highest priority are scheduled first to be sent to network.
+    HIGH_PRIORITY_SCHEDULER,
+    //! Priority with reservation scheduler: guarantee each DataWriter's minimum reservation of throughput.
+    //! Samples not fitting the reservation are scheduled by priority.
+    PRIORITY_WITH_RESERVATION_SCHEDULER
+};
+
+extern const char* const DEFAULT_FLOW_CONTROLLER_NAME;
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/rtps/flowcontrol/FlowControllerDefs.hpp]: ✅

File: include/fastdds/rtps/flowcontrol/FlowControllerDefs.hpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <fastrtps/fastrtps_dll.h>
+#include <cstdint>
+
+namespace eprosima {
+namespace fastdds {
+namespace rtps {
+
+/*!
+ * Supported scheduler plans by a flow controller.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/rtps/flowcontrol/FlowControllerDefs.hpp]: ✅

File: include/fastdds/rtps/flowcontrol/FlowControllerDefs.hpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * A flow controller's scheduler plan takes the decision of which samples are the next ones to be sent to the network.
+ * Fast-DDS flow controller supports several scheduler plans listed in this enumeration.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/rtps/flowcontrol/FlowControllerDefs.hpp]: ✅

File: include/fastdds/rtps/flowcontrol/FlowControllerDescriptor.hpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ * Configuration values for creating flow controllers.
+ *
+ * This descriptor is used to define the configuration applied in the creation of a flow controller.
+ */
+struct FlowControllerDescriptor
+{
+    //! Name of the flow controller.
+    const char* name = nullptr;
+
+    //! Scheduler plan used by the flow controller.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/rtps/flowcontrol/FlowControllerDescriptor.hpp]: ✅

File: include/fastdds/rtps/flowcontrol/FlowControllerDescriptor.hpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    FlowControllerScheduler scheduler = FlowControllerScheduler::FIFO_SCHEDULER;
+
+    //! Maximum number of bytes to be sent to network per period.
+    //!
+    //! Range of bytes: [1, 2147483647];
+    //! 0 value means no limit.
+    //! Default value: 0
+    int32_t max_bytes_per_period = 0;
+
+    //! Period time in milliseconds.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/rtps/flowcontrol/FlowControllerDescriptor.hpp:47]: ✅

File: include/fastdds/rtps/flowcontrol/FlowControllerDefs.hpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ */
+enum class FlowControllerScheduler : int32_t
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/rtps/flowcontrol/FlowControllerDefs.hpp]: ✅

**Review by rsanchez15 — APPROVED** [review]
Failed tests are not related to this PR.
Approved pending include header files style resolution.

File: include/fastdds/rtps/flowcontrol/FlowControllerDescriptor.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/rtps/flowcontrol/FlowControllerDescriptor.hpp]: As we do not have a rule yet, we can admit it by now.

**Pull request merged**